tests/bsim/bluetooth/mesh/src/test_suspend.c

```diff
@@ -0,0 +1,363 @@
+					  &status);
+	if (err || status) {
+		FAIL("Mod pub set failed (err %d, status %u)", err, status);
+	}
+
+	/* Wait until node has published before suspending. */
+	for (int i = 0; i < NUM_PUB; i++) {
+		err = k_sem_take(&publish_sem, K_SECONDS(30));
+		if (err) {
+			FAIL("Pub timed out");
+		}
+	}
+
+	/* TODO remove this sleep once suspend/resume works properly */
+	k_sleep(K_MSEC(100));
+	err = bt_mesh_suspend();
+	if (err) {
+		FAIL("Mesh suspend failed (err %d)", err);
+	}
+	suspended = true;
+	LOG_INF("Mesh suspended.");
+
+	k_sleep(K_SECONDS(SUSPEND_DURATION));
+
+	err = bt_mesh_resume();
+	if (err) {
+		FAIL("Mesh resume failed (err %d)", err);
+	}
+	suspended = false;
+	LOG_INF("Mesh resumed.");
+
+	for (int i = 0; i < NUM_PUB; i++) {
+		err = k_sem_take(&publish_sem, K_SECONDS(30));
+		if (err) {
+			FAIL("Pub timed out");
+		}
+	}
+
+	/* Allow publishing to finish before suspending. */
+	k_sleep(K_MSEC(100));
+	ASSERT_OK(bt_mesh_suspend());
+
+	PASS();
+}
+
+static void test_rx_suspend_resume(void)
+{
+	bt_mesh_test_cfg_set(NULL, WAIT_TIME);
+	bt_mesh_device_setup(&prov, &rx_comp);
+	provision_and_configure(rx_cfg, 1);
+
+	k_sem_init(&publish_sem, 0, 1);
+	int err;
+
+	for (int i = 0; i < NUM_PUB*2; i++) {
+		err = k_sem_take(&publish_sem, K_SECONDS(30));
+		if (err) {
+			FAIL("Receiver timed out");
+		}
+	}
+
+	PASS();
+}
+
+static void test_tx_suspend_disable_resume(void)
+{
+	bt_mesh_test_cfg_set(NULL, WAIT_TIME);
+	bt_mesh_device_setup(&prov, &tx_comp);
+	provision_and_configure(tx_cfg, 0);
+
+	k_sem_init(&publish_sem, 0, 1);
+	suspended = false;
+	uint8_t status;
+	int err;
+
+	err = bt_mesh_cfg_cli_mod_pub_set(0, tx_cfg.addr, tx_cfg.addr, TEST_MODEL_ID_1, &pub_params,
+					  &status);
+	if (err || status) {
+		FAIL("Mod pub set failed (err %d, status %u)", err, status);
+	}
+
+	/* Wait until node has published before suspending. */
+	for (int i = 0; i < NUM_PUB; i++) {
+		err = k_sem_take(&publish_sem, K_SECONDS(30));
+		if (err) {
+			FAIL("Pub timed out");
+		}
+	}
+
+	/* TODO remove this sleep once suspend/resume works properly */
+	k_sleep(K_MSEC(100));
+	err = bt_mesh_suspend();
+	if (err) {
+		FAIL("Mesh suspend failed (err %d)", err);
+	}
+	suspended = true;
+	LOG_INF("Mesh suspended.");
+
+	err = bt_disable();
+	if (err) {
+		FAIL("Bluetooth disable failed (err %d)", err);
+	}
+	LOG_INF("Bluetooth disabled.");
+
+	k_sleep(K_SECONDS(SUSPEND_DURATION));
+
+	err = bt_enable(NULL);
+	if (err) {
+		FAIL("Bluetooth init failed (err %d)", err);
+	}
+	LOG_INF("Bluetooth enabled.");
+
+	err = bt_mesh_resume();
+	if (err) {
+		FAIL("Mesh resume failed (err %d)", err);
+	}
+	suspended = false;
+	LOG_INF("Mesh resumed.");
+
+	for (int i = 0; i < NUM_PUB; i++) {
+		err = k_sem_take(&publish_sem, K_SECONDS(30));
+		if (err) {
+			FAIL("Pub timed out");
+		}
+	}
+
+	/* Allow publishing to finish before suspending. */
+	k_sleep(K_MSEC(100));
```

Comment:
Have you find the reason why it was failing without this delay?

Reply:
I have tried debugging quite a bit, and it seems that without the delay, the advertising thread goes into a spinlock which prevents further publication. I think when suspending the [advertising work](https://github.com/zephyrproject-rtos/zephyr/blob/e13912d92323c6ba94665c898408a18ef4cd7f3c/subsys/bluetooth/mesh/adv_ext.c#L65) immediately after publication (without the delay) it get preempted somewhere [here](https://github.com/zephyrproject-rtos/zephyr/blob/e13912d92323c6ba94665c898408a18ef4cd7f3c/subsys/bluetooth/mesh/adv_ext.c#L287-L289). 

I have a figure comparing the log output without (left) and with (right) the delay:
![image](https://github.com/zephyrproject-rtos/zephyr/assets/91412738/1e328274-7a60-4e16-83c3-ccba12ec769c)
 It may seem that some functionality in the access layer is disturbed by suspending immediately after a publication.

Reply:
What if schedule the adv work right after stopping and removing adv instances if `ADV_FLAG_SCHEDULED` or `ADV_FLAG_SCHEDULE_PENDING` is set? Then it will try to send the buffer and fail, but `struct bt_mesh_send_cb.start` will be called, which will notify the access layer.

Reply:
Fixed after offline discussion